runtime-sdk/src/state.rs

```diff
@@ -0,0 +1,1545 @@
+        self.emit_event_raw(event.into_event_tag());
+    }
+
+    /// Emit a raw event.
+    pub fn emit_event_raw(&mut self, etag: EventTag) {
+        let events = self.events.entry(etag.key).or_default();
+        events.push(etag.value);
+    }
+
+    /// Emit an unconditional event.
+    ///
+    /// The only difference to regular events is that these are handled as a separate set.
+    pub fn emit_unconditional_event<E: Event>(&mut self, event: E) {
+        let etag = event.into_event_tag();
+        let events = self.unconditional_events.entry(etag.key).or_default();
+        events.push(etag.value);
+    }
+
+    /// Take all regular events accumulated in the current state.
+    pub fn take_events(&mut self) -> EventTags {
+        mem::take(&mut self.events)
+    }
+
+    /// Take all unconditional events accumulated in the current state.
+    pub fn take_unconditional_events(&mut self) -> EventTags {
+        mem::take(&mut self.unconditional_events)
+    }
+
+    /// Take all events accumulated in the current state and return the merged set.
+    pub fn take_all_events(&mut self) -> EventTags {
+        let mut events = self.take_events();
+        let unconditional_events = self.take_unconditional_events();
+
+        for (key, val) in unconditional_events {
+            let tag = events.entry(key).or_default();
+            tag.extend(val)
+        }
+
+        events
+    }
+
+    /// Store associated with the state.
+    ///
+    /// # Panics
+    ///
+    /// This method will panic if no store exists.
+    pub fn store(&mut self) -> &mut dyn Store {
+        self.store.as_mut().unwrap()
+    }
+
+    /// Whether the store associated with the state has any pending updates.
+    pub fn has_pending_store_updates(&self) -> bool {
+        self.store
+            .as_ref()
+            .map(|store| store.has_pending_updates())
+            .unwrap_or_default()
+    }
+
+    /// Random number generator.
+    pub fn rng(&mut self) -> &mut RootRng {
+        self.rng.as_mut().unwrap()
+    }
+
+    /// Disables the RNG by replacing the instance with an invalid RNG.
+    fn disable_rng(&mut self) {
+        if self.parent.is_none() {
+            // Allowing hiding on root state would prevent the RNG from ever being recovered.
+            panic!("cannot hide the RNG on root state");
+        }
+        if self.hidden_rng.is_some() {
+            return; // Parent RNG already hidden.
+        }
+
+        self.hidden_rng = mem::replace(&mut self.rng, Some(RootRng::invalid()));
+    }
+
+    /// Environment information.
+    pub fn env(&self) -> &Environment {
+        &self.env
+    }
+
+    /// Returns the nesting level of the current state.
+    pub fn level(&self) -> usize {
+        if let Some(ref parent) = self.parent {
+            parent.level() + 1
+        } else {
+            0
+        }
+    }
+}
+
+thread_local! {
+    static CURRENT: RefCell<Vec<State>> = RefCell::new(Vec::new());
```

Comment:
can you summarize what mechanisms use the state .parent chain and what mechanisms use the CURRENT vector?

Reply:
The `CURRENT` vector is there so one can do `CurrentState::enter` multiple times (each `enter` results in a new element in this vector and the last element is the active one), starting with a fresh root state. The `.parent` chain is then derived from one such root state and is what is actually used most of the time (e.g. in practical use in the SDK there are no nested enters). All mechanisms only look at the last state in the `CURRENT` vector which is only used to support nested enters.

Alternatively we could forbid nested `CurrentState::enter` (e.g. panic if entered again instead of stacking), but supporting it may be useful (e.g. at least in tests).

Reply:
what is nested enter used for? something like when you're in the middle of a transaction in a block, and you want to simulate something happening before the block started?

`.parent` chain is used for things like smart contracts invoking other methods in my reading, is that right

Reply:
In practice nested enter is not used outside tests. Besides what you said it can also specify a completely different state store.

The "`.parent` state chain" is used for everything currently (simulation, subcalls from contracts, regular tx execution, ...).

Reply:
ah completely different state store. thanks for the summary, resolving